pkg/cloud/cloud.go

```diff
@@ -1271,49 +1261,71 @@ func (c *cloud) AvailabilityZones(ctx context.Context) (map[string]struct{}, err
 	return zones, nil
 }
 
-func (c *cloud) validateVolumeSize(ctx context.Context, volumeID string, newSizeGiB int64) (int64, error) {
+func (c *cloud) validateModifyVolume(ctx context.Context, volumeID string, newSizeGiB int64, options *ModifyDiskOptions) (bool, int64, error) {
 	request := &ec2.DescribeVolumesInput{
 		VolumeIds: []*string{
 			aws.String(volumeID),
 		},
 	}
 	volume, err := c.getVolume(ctx, request)
 	if err != nil {
-		return 0, err
+		return true, 0, err
 	}
 
 	oldSizeGiB := aws.Int64Value(volume.Size)
 
-	latestMod, modFetchError := c.getLatestVolumeModification(ctx, volumeID)
+	latestMod, err := c.getLatestVolumeModification(ctx, volumeID)
+	if err != nil && !errors.Is(err, VolumeNotBeingModified) {
+		return true, oldSizeGiB, fmt.Errorf("error fetching volume modifications for %q: %w", volumeID, err)
+	}
 
-	if latestMod != nil && modFetchError == nil {
+	// latestMod can be nil if the volume has never been modified
+	if latestMod != nil {
 		state := aws.StringValue(latestMod.ModificationState)
 		if state == ec2.VolumeModificationStateModifying {
-			err = c.waitForVolumeSize(ctx, volumeID)
+			// If volume is already modifying, detour to waiting for it to modify
+			klog.V(5).InfoS("[Debug] Watching ongoing modification", "volumeID", volumeID)
+			err = c.waitForVolumeModification(ctx, volumeID)
 			if err != nil {
-				return oldSizeGiB, err
+				return true, oldSizeGiB, err
 			}
-			return c.checkDesiredSize(ctx, volumeID, newSizeGiB)
+			returnGiB, returnErr := c.checkDesiredState(ctx, volumeID, newSizeGiB, options)
+			return false, returnGiB, returnErr
+		} else if state == ec2.VolumeModificationStateOptimizing {
+			return true, 0, fmt.Errorf("volume %q in OPTIMIZING state, cannot currently modify", volumeID)
 		}
 	}
 
-	// if there was an error fetching volume modifications and it was anything other than VolumeNotBeingModified error
-	// that means we have an API problem.
-	if modFetchError != nil && !errors.Is(modFetchError, VolumeNotBeingModified) {
-		return oldSizeGiB, fmt.Errorf("error fetching volume modifications for %q: %w", volumeID, modFetchError)
+	// At this point, we know we are starting a new volume modification
+	// If we're asked to modify a volume to its current state, ignore the request and immediately return a success
+	needsModification := false
+
+	if newSizeGiB != 0 && oldSizeGiB < newSizeGiB {
```

Comment:
Can this be simplified to `if newSizeGiB > oldSizeGiB` ? if that statement is true, then `newSizeGiB` should always be greater than 0.

Reply:
Yes, fixed